File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def dfs(self, image: List[List[int]], sr: int, sc: int, init: int, color: int) -> None:
        if sr < 0 or sr >= len(image) or sc < 0 or sc >= len(image[0]) or image[sr][sc] != init: return
        image[sr][sc] = color
        self.dfs(image, sr + 1, sc, init, color)
        self.dfs(image, sr, sc + 1, init, color)
        self.dfs(image, sr - 1, sc, init, color)
        self.dfs(image, sr, sc - 1, init, color)
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        if image[sr][sc] == color: return image
        self.dfs(image, sr, sc, image[sr][sc], color)
        return image

File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    int directions[4][2] = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image[sr][sc] == color) return image;
        queue<array<int, 2>> queue;
        queue.push({sr, sc});
        int init = image[sr][sc];
        image[sr][sc] = color;
        while(!queue.empty()) {
            int size = queue.size();
            while(size--) {
                auto arr = queue.front();
                queue.pop();
                for(auto dir : directions) {
                    int x = dir[0] + arr[0];
                    int y = dir[1] + arr[1];
                    if(x < 0 || x >= image.size() || y < 0 || y >= image[0].size() || image[x][y] != init) continue;
                    else {
                        image[x][y] = color;
                        queue.push({x,y});
                    }
                }
            }
        }
        return image;
    }
};